src/ray/util/logging.cc

```diff
@@ -418,58 +499,84 @@ void RayLog::AddFatalLogCallbacks(
+                                              std::to_string(GetTid()),
+                                              errno,
+                                              strerror(errno));
   }
   if (is_enabled_) {
-    logging_provider_ = new LoggingProvider(
-        file_name, line_number, GetMappedSeverity(severity), expose_osstream_);
+    if (log_format_json_) {
+      if (!component_name_.empty()) {
+        context_osstream_ << ",\"" << kLogKeyComponent << "\":\"" << component_name_
+                          << "\"";
+      }
+      context_osstream_ << ",\"" << kLogKeyFilename << "\":\"" << ConstBasename(file_name)
+                        << "\",\"" << kLogKeyLineno << "\":" << line_number;
+    } else {
+      if (!component_name_.empty()) {
+        msg_osstream_ << "(" << component_name_ << ") ";
+      }
+      msg_osstream_ << ConstBasename(file_name) << ":" << line_number << ": ";
+    }
   }
 }
 
-std::ostream &RayLog::Stream() {
-  auto logging_provider = reinterpret_cast<LoggingProvider *>(logging_provider_);
-  // Before calling this function, user should check IsEnabled.
-  // When IsEnabled == false, logging_provider_ will be empty.
-  return logging_provider->stream();
-}
-
 bool RayLog::IsEnabled() const { return is_enabled_; }
 
 bool RayLog::IsFatal() const { return is_fatal_; }
 
-std::ostream &RayLog::ExposeStream() { return *expose_osstream_; }
-
 RayLog::~RayLog() {
-  if (logging_provider_ != nullptr) {
-    delete reinterpret_cast<LoggingProvider *>(logging_provider_);
-    logging_provider_ = nullptr;
-  }
-  if (expose_osstream_ != nullptr) {
+  if (IsFatal()) {
+    msg_osstream_ << "\n*** StackTrace Information ***\n" << ray::StackTrace();
+    expose_fatal_osstream_ << "\n*** StackTrace Information ***\n" << ray::StackTrace();
     for (const auto &callback : fatal_log_callbacks_) {
-      callback(EL_RAY_FATAL_CHECK_FAILED, expose_osstream_->str());
+      callback(EL_RAY_FATAL_CHECK_FAILED, expose_fatal_osstream_.str());
     }
   }
+
+  auto logger = spdlog::get(RayLog::GetLoggerName());
+  if (!logger) {
+    logger = DefaultStdErrLogger::Instance().GetDefaultLogger();
+  }
+  // NOTE(lingxuan.zlx): See more fmt by visiting https://github.com/fmtlib/fmt.
+  if (log_format_json_) {
+    logger->log(GetMappedSeverity(severity_),
+                /*fmt*/ ",\"{}\":\"{}\"{}",
+                kLogKeyMessage,
+                json_escape_string(msg_osstream_.str()),
+                context_osstream_.str());
+  } else {
+    logger->log(GetMappedSeverity(severity_),
+                /*fmt*/ "{}{}",
+                msg_osstream_.str(),
+                context_osstream_.str());
+  }
+  logger->flush();
+
   if (severity_ == RayLogLevel::FATAL) {
     std::_Exit(EXIT_FAILURE);
   }
 }
 
+RayLog &RayLog::WithField(const std::string &key, const std::string &value) {
+  if (log_format_json_) {
+    context_osstream_ << ",\"" << key << "\":\"" << json_escape_string(value) << "\"";
```

Comment:
It's possible for user to provide a same key multiple times, resulting in a json with duplicated keys. It's technically legal but I don't think we want it. To avoid this, maybe we can do this:

1. in RayLog members, remove context_osstream_ and instead keep a map<str, str> fields_
2. in WithField, add to that fields_ (overwriting existing keys)
3. in ~RayLog, format the map in json or text mode, a la
```
#include <nlohmann/json.hpp>
#include "absl/strings/str_join.h"

std::string formatMap(const std::map<std::string, std::string>& map, bool jsonMode) {
    if (jsonMode) {
        nlohmann::json j(map);
        return j.dump();
    } else {
        return absl::StrJoin(map, " ", absl::PairFormatter("="));
    }
}
```